### ISSUE TITLE
[Feature request]: Automatic updating of the Document-Outline / TOC section

### ISSUE THREAD

**Issue by timunie** [RickStrahl/MarkdownMonster#800]
Hi Rick, 

## What is the issue?
I really love how easily I can add a TOC to my documents. Anyway, I noticed that I am sometimes a bit too lazy and forget to update the TOC before publish an updated document. 

## What is the feature request?
It would be very convenient if MM updates the TOC just before the file gets saved. I think the current TOC can be identified by the added comments: 

```markdown
<!-- Start Document Outline -->

* [Table of content](#table-of-content)
* [Introduction](#introduction)
[...]

<!-- End Document Outline -->
```
I know that there might be cases where one does not want to have such a behavior, so I think if you want this feature there should be an option to opt this in our out. 

## Are there any alternatives?
I have a Script for my own docs which does create a TOC for me. But I cannot use this everywhere, so it is just a partial solution. 

Happy coding
Tim

**Comment by RickStrahl** [comment]
It's relatively easy to get this done, but I'm not sure if we should do this :smile: I'm weary to start adding automation tasks into the base behavior of MM and extra setting checks which have to follow the document/project hierachy etc.

I wonder if something like this would help:

* Have a way to create a snippet that automatically hooks up the OnBeforeDocumentSaved() in which this could be manually automated.

The events exist but the problem currently is there's no easy way to hook them up to fire unless you create an addin.

**Comment by timunie** [comment]
Hi Rick, 

> Have a way to create a snippet that automatically hooks up the OnBeforeDocumentSaved() in which this could be manually automated.

"manually automated" sounds like digitization in Germany. ... Joke ... If this is somewhere documented it would be a way for me for sure. I don't need the TOC to be in the preview window. 

> The events exist but the problem currently is there's no easy way to hook them up to fire unless you create an addin.

An Add-In would also be a good idea. I can try to understand what I need to do and then create an addin. I am sure you have some docs somewhere. I could share the Add-In then as an OpenSource project if you don't mind me.

>Or another idea:
> Have snippets have set up in such a way they can be fired from a selection of common events in addition to key combos or manual invocation.

I am open to any solution. Sounds also good. 

To sum it up: For me the Add-In way seems to be the best option as you said I can already use the `OnBeforeDocumentSaved`-Event. Let's see if I manage to write my first MM-Plugin 😃 

Happy coding
Tim

**Comment by timunie** [comment]
Update: Here is my draft Add-In. 
- Does only run if there is already a TOC
- Does update the TOC only if needed in order to be more performant
- Can be opt in or out

Source: https://github.com/timunie/MarkdownMonsterTocUpdater

Still to do: 
- [ ] Create an better Icon
- [ ] Create the needed docs

Happy coding
Tim

**Comment by timunie** [comment]
Update 2: 

For me the Addin does what I want. I can close this topic now, if you don't need it anymore. 

By the way: Do you want to share the Addin via your Addin-Manager? 

Happy coding
Tim

**Comment by RickStrahl** [comment]
Looks like this addin will do what you need.

I'll keep a todo item open on my list for a way for snippets to intercept events so this can be create more easily. Seems like overkill to have an addin for such a simple task, but alas this is what works for now...
